kore/test/Test/Kore/AllPath.hs

```diff
@@ -0,0 +1,347 @@
+        )
+        `equals_`
+        (MultiOr.MultiOr [])
+    , AllPath.unprovenNodes
+        (emptyExecutionGraph (AllPath.Goal 0)
+            & insNode (1, AllPath.GoalRem 1)
+            & insEdge (0, 1)
+            & insNode (2, AllPath.Proven)
+            & insEdge (0, 2)
+        )
+        `equals_`
+        (MultiOr.MultiOr [1])
+    ]
+
+-- | Simple program configurations for unit testing.
+data K = BorC | A | B | C | D | E | F | Bot
+    deriving (Eq, Ord, Show)
+
+matches :: K -> K -> Bool
+matches B BorC = True
+matches C BorC = True
+matches a b    = a == b
+
+instance EqualWithExplanation K where
+    compareWithExplanation = rawCompareWithExplanation
+    printWithExplanation = show
+
+type Goal = (K, K)
+
+type ProofState = AllPath.ProofState Goal
+
+type Rule = (K, K)
+
+ruleAC :: Rule
+ruleAC = (A, C)
+
+cyclicRule :: [Rule]
+cyclicRule = [(A, A)]
+
+ruleAB, ruleAF :: Rule
+ruleAB = (A, B)
+ruleAF = (A, F)
+
+ruleBC, ruleBD, ruleBF :: Rule
+ruleBC = (B, C)
+ruleBD = (B, D)
+ruleBF = (B, F)
+
+ruleCD, ruleCF :: Rule
+ruleCD = (C, D)
+ruleCF = (C, F)
+
+ruleDE :: Rule
+ruleDE = (D, E)
+
+ruleEF :: Rule
+ruleEF = (E, F)
+
+noRules :: [Rule]
+noRules = []
+
+oneRule :: [Rule]
+oneRule = [ruleAB]
+
+concurrentRules :: [Rule]
+concurrentRules = [ ruleAB, ruleAC ]
+
+differentLengthPaths :: [Rule]
+differentLengthPaths =
+    [ ruleAB
+    , ruleAF
+    , ruleBC
+    , ruleBD
+    , ruleBF
+    , ruleCD
+    , ruleCF
+    , ruleDE
+    , ruleEF
+    ]
+
+type Prim = AllPath.Prim Rule
+
+-- | The destination-removal rule for our unit test goal.
+removeDestination :: Monad m => Goal -> m Goal
+removeDestination (src, dst) =
+    return (src', dst)
+  where
+    src'
+      | src `matches` dst        = Bot
+      | B    <- dst, BorC <- src = C
+      | C    <- dst, BorC <- src = B
+      | otherwise  = src
+
+-- | The goal is trivially valid when the members are equal.
+triviallyValid :: Goal -> Bool
+triviallyValid (src, _) = src == Bot
+
+derivePar :: [Rule] -> Goal -> Strategy.TransitionT Rule m ProofState
+derivePar rules (src, dst) =
+    goals <|> goalRem
+  where
+    goalRem
+      | any Maybe.isJust applied = (pure . AllPath.GoalRem) (Bot, dst)
+      | otherwise = (pure . AllPath.GoalRem) (src, dst)
+    applied = applyRule <$> rules
+    applyRule rule@(from, to)
+      | src `matches` from = Just $ do
+        Transition.addRule rule
+        (pure . AllPath.Goal) (to, dst)
+      | otherwise   = Nothing
+    goals = Foldable.asum (Maybe.fromMaybe empty <$> applied)
+
+runTransitionRule :: Prim -> ProofState -> [(ProofState, Seq Rule)]
+runTransitionRule prim state =
+    (runIdentity . runTransitionT) (transitionRule prim state)
+
+-- | 'AllPath.transitionRule' instantiated with our unit test rules.
+transitionRule
+    :: Prim
+    -> ProofState
+    -> Strategy.TransitionT Rule Identity ProofState
+transitionRule =
+    AllPath.transitionRule
+        removeDestination
+        triviallyValid
+        derivePar
+
```

Comment:
I'd be tempted to define things like `removeDestination` and `triviallyValid` after this definition. I'd forgotten what role they played until I saw how they were used.

That is, `where`-style rather than `let`-style.

I might actually recommend putting the `test_*` functions before the supporting definitions, assuming that people will want to read them first.